#Define the success of companies by calculating the company value


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**We firstly combine these three tables based on the crunchbase ID:**

“Acquisitions”

“Investment Round”

“Companies Crunchbase”

**name it as crunchbaseID_combined as below**

In [ ]:
import pandas as pd
data = pd.read_csv('./drive/MyDrive/crunchbaseID_combined.csv')

# Here we selected the data only for calculating company valuation
selected_data = data[['CrunchbaseID', 'Valuation At Ipo (in Usd)',
       'Funding Type','Money Raised Currency (in Usd)', 'Announced Date', 'Funding Stage',
       'Pre Money Valuation Currency (in Usd)', 'Equity Only Funding',
       'Price (in Usd)', 'A_Acquisition_Type']]

In [ ]:
# Convert the 'Announced Date' to datetime type
selected_data['Announced Date'] = pd.to_datetime(data['Announced Date'])

<ipython-input-4-ecc6f9316b04>:2: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  selected_data['Announced Date'] = pd.to_datetime(data['Announced Date'])
<ipython-input-4-ecc6f9316b04>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_data['Announced Date'] = pd.to_datetime(data['Announced Date'])


In [ ]:
# We sorted the data by the ascending order for 'CrunchbaseID'
df_sorted = selected_data.sort_values(by='CrunchbaseID')
# df_sorted.head(10)

In [ ]:
# We sorted the data by the ascending order for 'Announced Date' for each crunchbae
df_sorted1 = df_sorted.groupby('CrunchbaseID').apply(lambda x: x.sort_values(by='Announced Date')).reset_index(drop=True)
#df_sorted1.head(50)

In [ ]:
# We calculated the 'Post Money Valuation' by adding the 'Pre Money Valuation Currency (in Usd)' and 'Money Raised Currency (in Usd)' for each row
# post money is calculated only if both values are available
df_sorted1['Post Money Valuation Currency (in Usd)'] = df_sorted1['Pre Money Valuation Currency (in Usd)'] + df_sorted1['Money Raised Currency (in Usd)']
df_sorted1.head(10)

CrunchbaseID  Valuation At Ipo (in Usd)      Funding Type  \
0             1                        NaN             Grant   
1             2                        NaN              Seed   
2             3                        NaN               NaN   
3             4                        NaN               NaN   
4             5                        NaN               NaN   
5             6                        NaN              Seed   
6             7                        NaN               NaN   
7             8                        NaN  Convertible Note   
8             8                        NaN  Convertible Note   
9             9                        NaN               NaN   

   Money Raised Currency (in Usd) Announced Date Funding Stage  \
0                       2500000.0     2014-06-17           NaN   
1                        536251.0     2016-11-15          Seed   
2                             NaN            NaT           NaN   
3                             NaN            NaT           NaN   
4                             NaN            NaT           NaN   
5                             NaN     2016-06-01          Seed   
6                             NaN            NaT           NaN   
7                       1121000.0     2017-03-09           NaN   
8                        810500.0     2019-07-01           NaN   
9                             NaN            NaT           NaN   

   Pre Money Valuation Currency (in Usd) Equity Only Funding  Price (in Usd)  \
0                                    NaN                  No             NaN   
1                                    NaN                 Yes             NaN   
2                                    NaN                 NaN             NaN   
3                                    NaN                 NaN             NaN   
4                                    NaN                 NaN             NaN   
5                                    NaN                 Yes             NaN   
6                                    NaN                 NaN             NaN   
7                              5000000.0                  No             NaN   
8                                    NaN                  No             NaN   
9                                    NaN                 NaN             NaN   

  A_Acquisition_Type  Post Money Valuation Currency (in Usd)  
0                NaN                                     NaN  
1                NaN                                     NaN  
2                NaN                                     NaN  
3                NaN                                     NaN  
4                NaN                                     NaN  
5                NaN                                     NaN  
6                NaN                                     NaN  
7                NaN                               6121000.0  
8                NaN                                     NaN  
9                NaN                                     NaN

In [ ]:
# We only use the proxy method for post money estimated valuation when 'Equity Only Funding' == Yes
df_sorted1.loc[df_sorted1['Equity Only Funding'] == 'Yes', 'Post Money Estimated Valuation Currency (in Usd)'] = df_sorted1['Money Raised Currency (in Usd)'] / 0.2
df_sorted1.head(10)

CrunchbaseID  Valuation At Ipo (in Usd)      Funding Type  \
0             1                        NaN             Grant   
1             2                        NaN              Seed   
2             3                        NaN               NaN   
3             4                        NaN               NaN   
4             5                        NaN               NaN   
5             6                        NaN              Seed   
6             7                        NaN               NaN   
7             8                        NaN  Convertible Note   
8             8                        NaN  Convertible Note   
9             9                        NaN               NaN   

   Money Raised Currency (in Usd) Announced Date Funding Stage  \
0                       2500000.0     2014-06-17           NaN   
1                        536251.0     2016-11-15          Seed   
2                             NaN            NaT           NaN   
3                             NaN            NaT           NaN   
4                             NaN            NaT           NaN   
5                             NaN     2016-06-01          Seed   
6                             NaN            NaT           NaN   
7                       1121000.0     2017-03-09           NaN   
8                        810500.0     2019-07-01           NaN   
9                             NaN            NaT           NaN   

   Pre Money Valuation Currency (in Usd) Equity Only Funding  Price (in Usd)  \
0                                    NaN                  No             NaN   
1                                    NaN                 Yes             NaN   
2                                    NaN                 NaN             NaN   
3                                    NaN                 NaN             NaN   
4                                    NaN                 NaN             NaN   
5                                    NaN                 Yes             NaN   
6                                    NaN                 NaN             NaN   
7                              5000000.0                  No             NaN   
8                                    NaN                  No             NaN   
9                                    NaN                 NaN             NaN   

  A_Acquisition_Type  Post Money Valuation Currency (in Usd)  \
0                NaN                                     NaN   
1                NaN                                     NaN   
2                NaN                                     NaN   
3                NaN                                     NaN   
4                NaN                                     NaN   
5                NaN                                     NaN   
6                NaN                                     NaN   
7                NaN                               6121000.0   
8                NaN                                     NaN   
9                NaN                                     NaN   

   Post Money Estimated Valuation Currency (in Usd)  
0                                               NaN  
1                                         2681255.0  
2                                               NaN  
3                                               NaN  
4                                               NaN  
5                                               NaN  
6                                               NaN  
7                                               NaN  
8                                               NaN  
9                                               NaN

## Company value calculation


Through literature review, we have discovered that the primary focus in research measuring a company's success is the Acquisition Price, denoted as '**Price (in USD)**' in our dataset. Subsequently, we opted for the highest value among all acquisitions for each company in our dataset.

Following this, the Valuation at IPO emerges as the second most significant factor, represented by '**Valuation At Ipo (in USD)**' in our data.

In cases where both of these values are NaN, we computed the post-money valuation, utilizing the maximum value between **Post Money Valuation Currency** and **Post Money Estimated Valuation Currency**.

In [ ]:
def get_valuation(df):
    '''
    df: a selected dataframe with the same id
    '''
    # return the max price if exists
    max_price = df['Price (in Usd)'].max()
    if not np.isnan(max_price):
        return max_price
    # get unique ipo value for each id
    unique_ipo = df.drop_duplicates(subset=['CrunchbaseID'])['Valuation At Ipo (in Usd)'].iloc[0]
    if not np.isnan(unique_ipo):
        return unique_ipo
    # otherwise, return the maximum of all available valuation and -estimate_valuation
    return df[['Post Money Valuation Currency (in Usd)',
               'Post Money Estimated Valuation Currency (in Usd)']].max().max()

In [ ]:
# Assuming df_sorted1 is your DataFrame
import numpy as np
df = df_sorted1.groupby('CrunchbaseID').apply(get_valuation).reset_index()
df.columns = ['CrunchbaseID', 'company_value']

In [ ]:
df.head(5)

CrunchbaseID  company_value
0             1            NaN
1             2      2681255.0
2             3            NaN
3             4            NaN
4             5            NaN

In [ ]:
df.shape

(4827, 2)

In [ ]:
df.notna().sum()

CrunchbaseID     4827
company_value    1683
dtype: int64

Here, we have only 1683 companies with company  valuations, while 3144 companies fall into the unknown category. For the data in the unknown category, we cannot simply classify them as successful or unsuccessful companies; we are unsure how to handle this data. Additionally, for those companies with valuations, we need to consider which ones are deemed successful. However, we believe that this cannot be simply categorized as a binary classification problem (success and failure) because we have the exact valuations for each company. We also need to consider that companies with higher valuations are more likely to be successful. We are also uncertain about how to manage this aspect.

In [ ]:
df1 = df[df['company_value']>= 10000000]
df1.shape

(968, 2)

In [ ]:
df2 = df[df['company_value']>= 100000000]
df2.shape

(304, 2)

In [ ]:
df3 = df[df['company_value']>= 1000000000]
df3.shape

(50, 2)

In [ ]:
50/1683

0.029708853238265002

Among the companies with valuations of 1 billion dollars or more, we've only identified 50 such companies, with successful ones accounting for only about 3%.

For companies valued at or above 100 million, we found only 304 such companies, which constitutes a relatively small sample size for the application of our model in the future.

Additionally, for companies valued at or above 10 million, we identified 968 such companies, representing a relatively fair sample size for the application of our model in the future.

We will further evaluate which threshold makes more sense for assessing the success of a company.

In [ ]:
# Assuming df is your DataFrame with 'CrunchbaseID' and 'company_value' columns
not_null_ids = df.loc[df['company_value'].notnull(), 'CrunchbaseID'].tolist()
Companies_Crunchbase_match = pd.read_excel('./drive/MyDrive/Crunchbase Company Match.xlsx')
Companies_Crunchbase_match.head(10)

/usr/local/lib/python3.10/dist-packages/openpyxl/styles/stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


CrunchbaseMatchID  CrunchbaseID  CompanyID
0                  1             1          8
1                  2             2        105
2                  3             3        111
3                  4             4        131
4                  5             5        152
5                  6             6        170
6                  7             7        224
7                  8             8        235
8                  9             9        239
9                 10            10        263

In [ ]:
# Filter rows based on 'CrunchbaseID' existence in not_null_ids
filtered_df = Companies_Crunchbase_match[Companies_Crunchbase_match['CrunchbaseID'].isin(not_null_ids)]

# Display the new DataFrame
filtered_df.head(5)

CrunchbaseMatchID  CrunchbaseID  CompanyID
1                   2             2        105
7                   8             8        235
11                 12            12        276
12                 13            13        312
18                 19            19     144247

In [ ]:
import pandas as pd

# Assuming both DataFrames are defined as df and filtered_df
merged_df = filtered_df.merge(df, on='CrunchbaseID', how='left')

# Display the first 5 rows of the merged DataFrame
print(merged_df.head(5))


   CrunchbaseMatchID  CrunchbaseID  CompanyID  company_value
0                  2             2        105      2681255.0
1                  8             8        235      6121000.0
2                 12            12        276      2802345.0
3                 13            13        312     20632510.0
4                 19            19     144247      3381050.0


In [ ]:
# Save filtered_df as a CSV file
merged_df.to_csv('./drive/MyDrive/Filtered_crunchbaseID.csv', index=False)

# Data filtering and preprocessing

After finding out the companies with company valuations, we search the ’career’ table to ensure it contains only the work history of founders whose companies have an established company value.

##Create a new table named career_founders

In [ ]:
#Here we will merge useful information in the company table to the crunchbase company we want to study
company = pd.read_excel('./drive/MyDrive/Capstone/Define success/Company.xlsx')
founder_information = merged_df.merge(company, on='companyID', how='left')

In [ ]:
# We import the career table
career = pd.read_excel('./drive/MyDrive/Capstone/Define success/Career.xlsx')

/usr/local/lib/python3.10/dist-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


In [ ]:
# Count the unique FounderIDs
unique_founder_count = founder_information['FounderID'].nunique()

# Print the result
print(f"Number of unique FounderIDs: {unique_founder_count}")


Number of unique FounderIDs: 1692


**Sort each founder's work history by time**

In [ ]:
# Convert the 'Announced Date' to datetime type
career['Start Date'] = pd.to_datetime(career['Start Date'])
# We sorted the data by the ascending order for 'Announced Date' for each crunchbae
career = career.sort_values(by='FounderID')
career = career.groupby('FounderID').apply(lambda x: x.sort_values(by='Start Date')).reset_index(drop=True)

### **Merge the two tables so that career table contains only the work history of the founders that have founded companies that has existing company value**

In [ ]:
# Merge 'career' and 'founder_information' tables
merged_df = pd.merge(career, founder_information[['FounderID', 'CompanyID','company_value']], on='FounderID', how='inner')
# Rename the columns to avoid confusion
merged_df = merged_df.rename(columns={'CompanyID_x': 'CompanyID', 'CompanyID_y': 'Founder_CompanyID'})
merged_df_sorted = merged_df.sort_values(by='FounderID')

**Add the company value about the companies that founded by the founders to the table**

In [ ]:
# Function to map company_value to corresponding companyID
def map_company_value(row):
    if row['CompanyID'] == row['Founder_CompanyID']:
        return row['company_value']
    else:
        return pd.NA

# Apply the function to create a new column 'mapped_company_value'
merged_df_sorted['founded_company_value'] = merged_df.apply(map_company_value, axis=1)

**Drop the columns that are repeated**

In [ ]:
career_founders = merged_df_sorted.drop(['Founder_CompanyID', 'company_value'], axis=1)

**Drop the duplicated rows contains in the merged table**

This is because if the founder_id is related to two or more companies in the founder_information table, it will create repeated rows.

In [ ]:
career_founders = career_founders.drop_duplicates()

**Aggregated the duplicated rows has the different company value**

This is because if the founder_id is related to two or more companies in the founder_information table, it will create repeated rows for the founded companies, and one company value is NAN, and one has solid values, we just want the solid values. Here we achieve this by fill out the NA value with solid value, then drop the duplicated rows.

In [ ]:
# Define a custom aggregation function
import numpy as np
def aggregate_company_value(group):
    non_null_values = group['founded_company_value'].dropna()
    if non_null_values.empty:
        group['founded_company_value'] = np.nan
    else:
        group['founded_company_value'] = non_null_values.iloc[0]
    return group

# Apply the custom aggregation function to aggregate the data
career_founders_combined = career_founders.groupby(['FounderID', 'CompanyID']).apply(aggregate_company_value).reset_index(drop=True)
#career_founders_combined

In [ ]:
career_founders_combined = career_founders_combined.drop_duplicates()

**Repeat: sort each founder's work history by time**

In [ ]:
career_founders_combined = career_founders_combined.sort_values(by='FounderID')
career_founders_combined = career_founders_combined.groupby('FounderID').apply(lambda x: x.sort_values(by='Start Date')).reset_index(drop=True)

#career_founders_combined.head(10)

In [ ]:
career_founders_combined.shape

(13872, 15)

In [ ]:
career_founders_combined.head(5)

CareerID  FounderID  CompanyID            JobTitle        DateRange  \
0         8          1      53036      Vise President        2005–2009   
1         7          1      53036  Executive Director        2007–2009   
2    216950          1      19921             Founder  01/2016-Present   
3    216951          1      19879              Member     2016-Present   
4         4          1      54711              Member     2016–Present   

  Start Date   End Date  Duration (years)                    Location  \
0 2005-03-01 2009-09-01               4.5                         NaN   
1 2007-03-01 2009-09-01               2.5  Moscow, Russian Federation   
2 2016-01-01        NaT               6.4                      Cyprus   
3 2016-03-01        NaT               6.2                      Berlin   
4 2016-03-01        NaT               5.4                         NaN   

                                         Description Created Date  Relevant  \
0                                                NaN   2021-02-19     False   
1                                                NaN   2021-02-19     False   
2  We are the largest global \ntravel mobility ma...   2022-05-05      True   
3                                                NaN   2022-05-05     False   
4            Europe’s Largest Technology Marketplace   2021-02-19     False   

   founded_company_value Headquarters Location Number Of Employees  
0                    NaN                   NaN                 NaN  
1                    NaN                   NaN                 NaN  
2             22450000.0   European Union (EU)                 NaN  
3                    NaN                   NaN                 NaN  
4                    NaN                   NaN                 NaN

In [ ]:
# Count the unique FounderIDs
unique_founder_count = career_founders_combined['FounderID'].nunique()

# Print the result
print(f"Number of unique FounderIDs: {unique_founder_count}")


Number of unique FounderIDs: 1692


##Next step: we only want the employment history before the founder founded their companies, for people that founded many companies, we can split his experience as serval times.

**Firstly, we count the number of companies that each founder that founded.**

In [ ]:
result = founder_information.groupby('FounderID')['company_value'].count()
len(result)

1692

### For founders who founded only one company and who founded more than one company, we should treat differently.


For founders that only founded one company, we only consider his employment history before this specific company. For founders that founded more than one company, we need to split his employment history to different times.

For founders that founded more than one company: **Firstly, let us clean up the data, to make sure that the founders indeed have founded different companies.**

For example, although FounderID=399 has records of multiple business startup, it is indeed the same company, we only consider the employment history before he first founded this company, i.e., we treat him as an entrepreneur who has only founded one company as in the previous case. We can find the FounderID of them in the table 'founder_information'


In [ ]:
founder_399_rows = founder_information[founder_information['FounderID'] == 399]
founder_399_rows

company_value  CrunchbaseID Company Name         Industries  \
848      100000000          1942        Sendy  Computer Software   
855      100000000          1942        Sendy  Computer Software   
856      100000000          1942        Sendy  Computer Software   

    Headquarters Location       Founded Date Number Of Employees  CareerID  \
848               Nairobi  1/1/2014 00:00:00              51-200      2224   
855               Nairobi  1/1/2014 00:00:00              51-200    263683   
856               Nairobi  1/1/2014 00:00:00              51-200    263681   

     FounderID  CompanyID           JobTitle          DateRange  \
848        399      43516   Co-Founder & COO  Jan 2015–Dec 2018   
855        399      43516   Co-Founder & CFO      1/2019-7/2022   
856        399      43516  Co-Founder & CIRO     7/2022-Present   

            Start Date            End Date  Duration (years)        Location  \
848  1/1/2015 00:00:00  1/12/2018 00:00:00               3.9  Nairobi, Kenya   
855  1/1/2019 00:00:00   1/7/2022 00:00:00               3.5         Nairobi   
856  1/7/2022 00:00:00                 NaN               0.3  Nairobi, Kenya   

    Description        Created Date  Relevant  
848         NaN  18/2/2021 00:00:00         1  
855         NaN  6/10/2022 00:00:00         1  
856         NaN  6/10/2022 00:00:00         1

**Secondly, we can drop the duplicate rows for the same Founder ID with same company ID.**

In [ ]:
founder_information = founder_information.sort_values(by='FounderID')
founder_information = founder_information.groupby('FounderID').apply(lambda x: x.sort_values(by='Start Date')).reset_index(drop=True)
founder_information.shape

(1925, 19)

In [ ]:
# Delete the duplicates and keep the first row
founder_information_dropped = founder_information.drop_duplicates(subset=['FounderID', 'CompanyID'], keep='first')
founder_information_dropped = founder_information_dropped.sort_values(by='FounderID')
founder_information_dropped = founder_information_dropped.groupby('FounderID').apply(lambda x: x.sort_values(by='Start Date')).reset_index(drop=True)
founder_information_dropped.shape

(1801, 19)

In [ ]:
founder_399_rows = founder_information_dropped[founder_information_dropped['FounderID'] == 399]
founder_399_rows

company_value  CrunchbaseID Company Name         Industries  \
24      100000000          1942        Sendy  Computer Software   

   Headquarters Location       Founded Date Number Of Employees  CareerID  \
24               Nairobi  1/1/2014 00:00:00              51-200      2224   

    FounderID  CompanyID          JobTitle          DateRange  \
24        399      43516  Co-Founder & COO  Jan 2015–Dec 2018   

           Start Date            End Date  Duration (years)        Location  \
24  1/1/2015 00:00:00  1/12/2018 00:00:00               3.9  Nairobi, Kenya   

   Description        Created Date  Relevant  
24         NaN  18/2/2021 00:00:00         1

In [ ]:
# Assuming you have the 'result' Series
founder_ids_with_result_1 = result[result == 1].index.tolist()

# Print the founder IDs
len(founder_ids_with_result_1)

1591

### For founders that only founded one company, we only consider his employment history before this specific company.


In [ ]:
# Filter the DataFrame based on the founder IDs
career_founders_with_1 = career_founders_combined[career_founders_combined['FounderID'].isin(founder_ids_with_result_1)]
career_founders_with_1 = career_founders_with_1.sort_values(by='FounderID')
career_founders_with_1 = career_founders_with_1 .groupby('FounderID').apply(lambda x: x.sort_values(by='Start Date')).reset_index(drop=True)

In [ ]:
career_founders_with_1.shape

(12863, 15)

In [ ]:
# Define a function to filter rows
def filter_rows(df):
    # Find the index of the first row with a non-solid value (NaN) in 'founded_company_value'
    first_invalid_index = df['founded_company_value'].first_valid_index()
    # Filter the DataFrame up to and including the first non-solid value
    return df.loc[:first_invalid_index]

# Apply the function to each group and concatenate the results
career_founders_with_1 = pd.concat([filter_rows(group) for _, group in career_founders_with_1.groupby('FounderID')])

# Print the filtered DataFrame
career_founders_with_1.head(5)

CareerID  FounderID  CompanyID  \
0         8          1      53036   
1         7          1      53036   
2    216950          1      19921   
6     80467          3      27310   
7     80414          3      39824   

                                        JobTitle          DateRange  \
0                                 Vise President          2005–2009   
1                             Executive Director          2007–2009   
2                                        Founder    01/2016-Present   
6                                Lead Accountant  Nov 2001–Apr 2006   
7  Manager of Finance and Accounting, CIS Region  May 2006–Oct 2008   

  Start Date   End Date  Duration (years)                    Location  \
0 2005-03-01 2009-09-01               4.5                         NaN   
1 2007-03-01 2009-09-01               2.5  Moscow, Russian Federation   
2 2016-01-01        NaT               6.4                      Cyprus   
6 2001-11-01 2006-04-01               4.4                         NaN   
7 2006-05-01 2008-10-01               2.4                         NaN   

                                         Description Created Date  Relevant  \
0                                                NaN   2021-02-19     False   
1                                                NaN   2021-02-19     False   
2  We are the largest global \ntravel mobility ma...   2022-05-05      True   
6                                                NaN   2021-03-10     False   
7                                                NaN   2021-03-10     False   

   founded_company_value Headquarters Location Number Of Employees  
0                    NaN                   NaN                 NaN  
1                    NaN                   NaN                 NaN  
2             22450000.0   European Union (EU)                 NaN  
6                    NaN                   NaN                 NaN  
7                    NaN                   NaN                 NaN

Here we create a new index for each founding experience. For each work experience before the founding experience we number the order by date, and the founding experience is the last index. (for example, for founder whose FounderID = 1, a total of three index [0,1,2], and index 2 corresponds to the company he founded the company)

In [ ]:
# Reset the index for each different founder ID starting from 0
career_founders_with_1['index'] = career_founders_with_1.groupby('FounderID').cumcount()
career_founders_with_1.reset_index(drop=True, inplace=True)
# Print the modified DataFrame
career_founders_with_1.head(5)

CareerID  FounderID  CompanyID  \
0         8          1      53036   
1         7          1      53036   
2    216950          1      19921   
3     80467          3      27310   
4     80414          3      39824   

                                        JobTitle          DateRange  \
0                                 Vise President          2005–2009   
1                             Executive Director          2007–2009   
2                                        Founder    01/2016-Present   
3                                Lead Accountant  Nov 2001–Apr 2006   
4  Manager of Finance and Accounting, CIS Region  May 2006–Oct 2008   

  Start Date   End Date  Duration (years)                    Location  \
0 2005-03-01 2009-09-01               4.5                         NaN   
1 2007-03-01 2009-09-01               2.5  Moscow, Russian Federation   
2 2016-01-01        NaT               6.4                      Cyprus   
3 2001-11-01 2006-04-01               4.4                         NaN   
4 2006-05-01 2008-10-01               2.4                         NaN   

                                         Description Created Date  Relevant  \
0                                                NaN   2021-02-19     False   
1                                                NaN   2021-02-19     False   
2  We are the largest global \ntravel mobility ma...   2022-05-05      True   
3                                                NaN   2021-03-10     False   
4                                                NaN   2021-03-10     False   

   founded_company_value Headquarters Location Number Of Employees  index  
0                    NaN                   NaN                 NaN      0  
1                    NaN                   NaN                 NaN      1  
2             22450000.0   European Union (EU)                 NaN      2  
3                    NaN                   NaN                 NaN      0  
4                    NaN                   NaN                 NaN      1

In [ ]:
career_founders_with_1.shape

(9724, 16)

### For founders that founded more than one companie, we split his founding history.


In [ ]:
# Assuming you have the 'result' Series
founder_ids_with_result_more_companies = result[result != 1].index.tolist()

# Print the founder IDs
len(founder_ids_with_result_more_companies)

101

In [ ]:
max(result)

3

In [ ]:
# Filter the DataFrame based on the founder IDs
career_founders_with_more_companies = career_founders_combined[career_founders_combined['FounderID'].isin(founder_ids_with_result_more_companies)]
career_founders_with_more_companies = career_founders_with_more_companies.sort_values(by='FounderID')
career_founders_with_more_companies = career_founders_with_more_companies .groupby('FounderID').apply(lambda x: x.sort_values(by='Start Date')).reset_index(drop=True)
career_founders_with_more_companies.reset_index(drop=True, inplace=True)
#career_founders_with_more_companies

In the case of a founder who founded multiple companies, we treat all of his experiences prior to the founding of a particular company as a single founding experience. For example, if an entrepreneur starts two companies, then all the work experience prior to starting the first company is considered as one startup experience. For the second company, the work experience prior to the founding of the second company (including the founding of the first company) is treated as another founding experience, i.e., a total of two founding experiences for this entrepreneur.

In [ ]:
def filter_rows(df):
    # Calculate the number of companies founded for each founder ID
    founded_company_indices = df['founded_company_value'].index[df['founded_company_value'].notnull()]
    founded_company_numbers = len(founded_company_indices)

    # Initialize an empty DataFrame to store concatenated records
    career_record = pd.DataFrame()

    # Concatenate records for each founded company
    for i in range(founded_company_numbers):
        valid_index = founded_company_indices[i]
        company_record = df.loc[:valid_index].copy()  # Make a copy to avoid SettingWithCopyWarning
        company_record['index'] = range(len(company_record))  # Add index for each career record
        career_record = pd.concat([career_record, company_record])

    return career_record

In [ ]:
# Group by 'FounderID'
grouped_founders = career_founders_with_more_companies.groupby('FounderID')

# Initialize an empty DataFrame to store the concatenated results
career_founders_with_more_companies_filtered = pd.DataFrame()

# Apply the function to each group and concatenate the results
for founder_id, group_df in grouped_founders:
    filtered_records = filter_rows(group_df)
    career_founders_with_more_companies_filtered = pd.concat([career_founders_with_more_companies_filtered, filtered_records])

# Now 'all_filtered_records' contains the filtered data for all company founders
career_founders_with_more_companies_filtered.head(10)

CareerID  FounderID  CompanyID                                 JobTitle  \
0    170623        254      87576           Vice-director CISV Summer Camp   
1    170624        254      87577             Corporate Immigration Intern   
2      1583        254      33806    Marketing and Communication Assistant   
3      1582        254      41556                       Project Consultant   
4      1581        254      30203                           Executive Team   
5      1580        254      11183  Fundraising and Communication Assistant   
6      1579        254      56403                         Co-Founder & COO   
0    170623        254      87576           Vice-director CISV Summer Camp   
1    170624        254      87577             Corporate Immigration Intern   
2      1583        254      33806    Marketing and Communication Assistant   

           DateRange Start Date   End Date  Duration (years)  \
0  May 2011–Sep 2011 2011-05-01 2011-09-01               0.3   
1  Sep 2013–Dec 2013 2013-09-01 2013-12-01               0.2   
2  Feb 2014–May 2014 2014-02-01 2014-05-01               0.2   
3  Nov 2014–Apr 2015 2014-11-01 2015-04-01               0.4   
4  Jan 2015–May 2015 2015-01-01 2015-05-01               0.3   
5  Oct 2015–Jan 2016 2015-10-01 2016-01-01               0.3   
6   Feb 2016–Present 2016-02-01        NaT               5.5   
0  May 2011–Sep 2011 2011-05-01 2011-09-01               0.3   
1  Sep 2013–Dec 2013 2013-09-01 2013-12-01               0.2   
2  Feb 2014–May 2014 2014-02-01 2014-05-01               0.2   

                    Location  \
0        Ferrara Area, Italy   
1          Siena Area, Italy   
2   Montpellier Area, France   
3     London, United Kingdom   
4     London, United Kingdom   
5  Cambridge, United Kingdom   
6     London, United Kingdom   
0        Ferrara Area, Italy   
1          Siena Area, Italy   
2   Montpellier Area, France   

                                         Description Created Date  Relevant  \
0                                                NaN   2021-07-23     False   
1                                                NaN   2021-07-23     False   
2                                                NaN   2021-02-18     False   
3  Consultancy project as required by Masters reg...   2021-02-18     False   
4                                                NaN   2021-02-18     False   
5                                                NaN   2021-02-18     False   
6  We are building Zazu: the helpful app to manag...   2021-02-18      True   
0                                                NaN   2021-07-23     False   
1                                                NaN   2021-07-23     False   
2                                                NaN   2021-02-18     False   

   founded_company_value                Headquarters Location  \
0                    NaN                                  NaN   
1                    NaN                                  NaN   
2                    NaN                                  NaN   
3                    NaN                                  NaN   
4                    NaN                               London   
5                    NaN                                  NaN   
6              7000000.0  Lusaka, NA - Zambia, Zambia, Africa   
0                    NaN                                  NaN   
1                    NaN                                  NaN   
2                    NaN                                  NaN   

  Number Of Employees  index  
0                 NaN      0  
1                 NaN      1  
2                 NaN      2  
3                 NaN      3  
4               11-50      4  
5                 NaN      5  
6               11-50      6  
0                 NaN      0  
1                 NaN      1  
2                 NaN      2

In [ ]:
career_founders_with_more_companies_filtered.shape

(1616, 16)

### Combine the two tables

In [ ]:
# Here we concat our two tables as our final data
data = pd.concat([career_founders_with_1, career_founders_with_more_companies_filtered], ignore_index=True)

In [ ]:
data.shape

(11340, 16)

In [ ]:
data.head(5)

CareerID  FounderID  CompanyID  \
0         8          1      53036   
1         7          1      53036   
2    216950          1      19921   
3     80467          3      27310   
4     80414          3      39824   

                                        JobTitle          DateRange  \
0                                 Vise President          2005–2009   
1                             Executive Director          2007–2009   
2                                        Founder    01/2016-Present   
3                                Lead Accountant  Nov 2001–Apr 2006   
4  Manager of Finance and Accounting, CIS Region  May 2006–Oct 2008   

  Start Date   End Date  Duration (years)                    Location  \
0 2005-03-01 2009-09-01               4.5                         NaN   
1 2007-03-01 2009-09-01               2.5  Moscow, Russian Federation   
2 2016-01-01        NaT               6.4                      Cyprus   
3 2001-11-01 2006-04-01               4.4                         NaN   
4 2006-05-01 2008-10-01               2.4                         NaN   

                                         Description Created Date  Relevant  \
0                                                NaN   2021-02-19     False   
1                                                NaN   2021-02-19     False   
2  We are the largest global \ntravel mobility ma...   2022-05-05      True   
3                                                NaN   2021-03-10     False   
4                                                NaN   2021-03-10     False   

   founded_company_value Headquarters Location Number Of Employees  index  
0                    NaN                   NaN                 NaN      0  
1                    NaN                   NaN                 NaN      1  
2             22450000.0   European Union (EU)                 NaN      2  
3                    NaN                   NaN                 NaN      0  
4                    NaN                   NaN                 NaN      1

The final 'data' represent the career records of the founders of companies with existing company value. This data will be used for both the Exploratory Data Analysis and our prediction model.
